src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinActivityConversionExtensions.cs

```diff
@@ -0,0 +1,219 @@
+                annotations,
+                attributeEnumerationState.Tags,
+                null,
+                null);
+        }
+
+        internal static string EncodeSpanId(ActivitySpanId spanId)
+        {
+            return spanId.ToHexString();
+        }
+
+        internal static long ToEpochMicroseconds(this DateTimeOffset timestamp)
+        {
+            return timestamp.ToUnixTimeMilliseconds() * 1000;
+        }
+
+        internal static long ToEpochMicroseconds(this DateTime utcDateTime)
+        {
+            const long TicksPerMicrosecond = TimeSpan.TicksPerMillisecond / 1000;
+            const long UnixEpochTicks = 621355968000000000; // = DateTimeOffset.FromUnixTimeMilliseconds(0).Ticks
+            const long UnixEpochMicroseconds = UnixEpochTicks / TicksPerMicrosecond;
+
+            // Truncate sub-microsecond precision before offsetting by the Unix Epoch to avoid
+            // the last digit being off by one for dates that result in negative Unix times
+            long microseconds = utcDateTime.Ticks / TicksPerMicrosecond;
+            return microseconds - UnixEpochMicroseconds;
+        }
+
+        private static string EncodeTraceId(ActivityTraceId traceId, bool useShortTraceIds)
+        {
+            var id = traceId.ToHexString();
+
+            if (id.Length > 16 && useShortTraceIds)
+            {
+                id = id.Substring(id.Length - 16, 16);
+            }
+
+            return id;
+        }
+
+        private static string ToActivityKind(Activity activity)
+        {
+            switch (activity.Kind)
+            {
+                case ActivityKind.Server:
+                    return "SERVER";
+                case ActivityKind.Producer:
+                    return "PRODUCER";
+                case ActivityKind.Consumer:
+                    return "CONSUMER";
+                default:
+                    return "CLIENT";
+            }
+        }
+
+        private static bool ProcessActivityEvents(ref PooledList<ZipkinAnnotation> annotations, ActivityEvent @event)
+        {
+            PooledList<ZipkinAnnotation>.Add(ref annotations, new ZipkinAnnotation(@event.Timestamp.ToEpochMicroseconds(), @event.Name));
+            return true;
+        }
+
+        private static bool ProcessTags(ref AttributeEnumerationState state, KeyValuePair<string, string> attribute)
+        {
+            string key = attribute.Key;
+            string strVal = attribute.Value?.ToString();
+
+            if (strVal != null)
+            {
+                if (RemoteEndpointServiceNameKeyResolutionDictionary.TryGetValue(key, out int priority)
+                    && (state.RemoteEndpointServiceName == null || priority < state.RemoteEndpointServiceNamePriority))
+                {
+                    state.RemoteEndpointServiceName = strVal;
+                    state.RemoteEndpointServiceNamePriority = priority;
+                }
+                else if (key == Resource.ServiceNameKey)
```

Comment:
Curious, wonder if "if-else" or a dict/hash look up would be more performant. Probably won't matter as it is not on the hot path.